feature/manager/src/main/java/com/kinandcarta/create/proxytoggle/manager/viewmodel/ProxyManagerViewModel.kt

```diff
@@ -0,0 +1,215 @@
+        updateUiState {
+            it.copy(portState = it.portState.copy(text = newTextFiltered))
+        }
+    }
+
+    fun onForceFocusExecuted() {
+        updateUiState {
+            it.copy(
+                addressState = it.addressState.copy(forceFocus = false),
+                portState = it.portState.copy(forceFocus = false)
+            )
+        }
+    }
+
+    private fun enableProxy() {
+        val address = uiState.value.addressState.text
+        val port = uiState.value.portState.text
+
+        updateErrors(ProxyManagerError.NoError)
+        updateUiState {
+            it.copy(
+                addressState = it.addressState.copy(forceFocus = false),
+                portState = it.portState.copy(forceFocus = false)
+            )
+        }
+        when {
+            !proxyValidator.isValidIP(address) -> {
+                viewModelScope.launch {
+                    delay(ERROR_DELAY)
+                    updateErrors(ProxyManagerError.InvalidAddress)
+                }
+            }
+            !proxyValidator.isValidPort(port) -> {
+                viewModelScope.launch {
+                    delay(ERROR_DELAY)
+                    updateErrors(ProxyManagerError.InvalidPort)
+                }
+            }
+            else -> {
+                deviceSettingsManager.enableProxy(Proxy(address, port))
+            }
+        }
+    }
+
+    private fun updateUiState(updateFunc: (UiState) -> UiState) {
+        _uiState.value = updateFunc(_uiState.value)
+    }
+
+    private fun updateErrors(errorState: ProxyManagerError) {
+        when (errorState) {
+            ProxyManagerError.InvalidAddress -> {
+                updateUiState {
+                    it.copy(
+                        addressState = it.addressState.copy(
+                            error = R.string.error_invalid_address,
+                            forceFocus = true
+                        )
+                    )
+                }
+            }
+            ProxyManagerError.InvalidPort -> {
+                updateUiState {
+                    it.copy(
+                        portState = it.portState.copy(
+                            error = R.string.error_invalid_port,
+                            forceFocus = true
+                        )
+                    )
+                }
+            }
+            ProxyManagerError.NoError -> {
+                updateUiState {
+                    it.copy(
+                        addressState = it.addressState.copy(error = null),
+                        portState = it.portState.copy(error = null)
+                    )
+                }
+            }
+        }
+    }
+
+    private fun getInitialAddressState(): TextFieldState {
+        return TextFieldState(
+            label = context.getString(R.string.hint_ip_address),
+            text = getLastUsedProxy()?.address ?: "",
+            keyboardOptions = getKeyboardOptions(KeyboardType.Uri, ImeAction.Next)
+        )
+    }
```

Comment:
Normally no Context should be involved in the ViewModel.

Reply:
Yep. Originally, I was passing the StringRes into the composables and using `text = stringResource(resId)` in there, but I felt like we should pass the exact text, ready-made from the viewmodel (as part of the movement of all logic in there). I can revert though, no problem.

Reply:
Done